File: js/src/message.ts
import { textDecoder, textEncoder, toUint8Array, type Bytes } from "./bytes.js";

export enum MsgType {
  Command = 1 << 0,
  Response = 1 << 1,
  NeedRsp = 1 << 2,
  Ping = 1 << 3,
  Pong = 1 << 4,
  NoSuchCmd = 1 << 5,
}

export interface RpcMessage {
  seq: number;
  type: number;
  cmd: string;
  data: Uint8Array;
  requestPayload?: Uint8Array;
}

const PAYLOAD_MIN_LEN = 4 + 2 + 1;

export function hasType(type: number, flag: MsgType): boolean {
  return (type & flag) !== 0;
}

export function serializeMessage(msg: RpcMessage): Uint8Array {
  const cmd = textEncoder.encode(msg.cmd);
  if (cmd.byteLength > 0xffff) {
    throw new RangeError("cmd length exceeds uint16");
  }

  const data = msg.requestPayload ?? msg.data;
  const payload = new Uint8Array(PAYLOAD_MIN_LEN + cmd.byteLength + data.byteLength);
  const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  let offset = 0;

  view.setUint32(offset, msg.seq >>> 0, true);
  offset += 4;
  view.setUint16(offset, cmd.byteLength, true);
  offset += 2;
  payload.set(cmd, offset);
  offset += cmd.byteLength;
  view.setUint8(offset, msg.type & 0xff);
  offset += 1;
  payload.set(data, offset);

  return payload;
}

export function deserializeMessage(payload: Bytes): RpcMessage | undefined {
  const bytes = toUint8Array(payload);
  if (bytes.byteLength < PAYLOAD_MIN_LEN) {
    return undefined;
  }

  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  let offset = 0;
  const seq = view.getUint32(offset, true);
  offset += 4;
  const cmdLen = view.getUint16(offset, true);
  offset += 2;
  if (offset + cmdLen + 1 > bytes.byteLength) {
    return undefined;
  }

  const cmd = textDecoder.decode(bytes.subarray(offset, offset + cmdLen));
  offset += cmdLen;
  const type = view.getUint8(offset);
  offset += 1;
  const data = bytes.slice(offset);

  return { seq, type, cmd, data };
}

File: python/rpc_core/message.py
from __future__ import annotations

from dataclasses import dataclass
from enum import IntFlag


class MsgType(IntFlag):
    COMMAND = 1 << 0
    RESPONSE = 1 << 1
    NEED_RSP = 1 << 2
    PING = 1 << 3
    PONG = 1 << 4
    NO_SUCH_CMD = 1 << 5


@dataclass
class RpcMessage:
    seq: int
    type: int
    cmd: str
    data: bytes
    request_payload: bytes | None = None


PAYLOAD_MIN_LEN = 4 + 2 + 1


def has_type(type_value: int, flag: MsgType) -> bool:
    return (type_value & int(flag)) != 0


def serialize_message(msg: RpcMessage) -> bytes:
    cmd = msg.cmd.encode("utf-8")
    if len(cmd) > 0xFFFF:
        raise ValueError("cmd length exceeds uint16")

    data = msg.request_payload if msg.request_payload is not None else msg.data
    payload = bytearray(PAYLOAD_MIN_LEN + len(cmd) + len(data))
    offset = 0
    payload[offset : offset + 4] = (msg.seq & 0xFFFFFFFF).to_bytes(4, "little")
    offset += 4
    payload[offset : offset + 2] = len(cmd).to_bytes(2, "little")
    offset += 2
    payload[offset : offset + len(cmd)] = cmd
    offset += len(cmd)
    payload[offset] = msg.type & 0xFF
    offset += 1
    payload[offset:] = data
    return bytes(payload)


def deserialize_message(payload: bytes | bytearray | memoryview) -> RpcMessage | None:
    data = bytes(payload)
    if len(data) < PAYLOAD_MIN_LEN:
        return None

    offset = 0
    seq = int.from_bytes(data[offset : offset + 4], "little")
    offset += 4
    cmd_len = int.from_bytes(data[offset : offset + 2], "little")
    offset += 2
    if offset + cmd_len + 1 > len(data):
        return None

    cmd = data[offset : offset + cmd_len].decode("utf-8")
    offset += cmd_len
    type_value = data[offset]
    offset += 1
    return RpcMessage(seq=seq, type=type_value, cmd=cmd, data=data[offset:])
